common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java

```diff
@@ -350,18 +350,22 @@ public void onSSLStateChange(int type, int val) {
 
     @Override
     @SuppressWarnings("unused") // used by NativeCrypto.SSLHandshakeCallbacks / new_session_callback
-    public boolean onNewSessionEstablished(long sslSessionNativePtr) {
+    public void onNewSessionEstablished(long sslSessionNativePtr) {
+        NativeCrypto.SSL_SESSION_up_ref(sslSessionNativePtr);
         SslSessionWrapper sessionWrapper;
         try {
-            sessionWrapper = SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession);
-        } catch (Throwable e) {
-            return false;
+            sessionWrapper =
+                SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession);
+        } catch (Exception ignored) {
+            // Failed constructing the wrapper. Its finalizer will not release the reference, so
+            // we need to here.
+            NativeCrypto.SSL_SESSION_free(sslSessionNativePtr);
```

Comment:
I think this works, since you moved the saving to the end, but this relies on, e.g., `SslSessionWrapper.newInstance` not doing anything after creating the `SslSessionWrapper`. Another option would be to say `SslSessionWrapper.newInstance` and `SslSessionWrapper.Impl()` do *not* take ownership of the reference. Instead, the `Impl` constructor just looks like:

```
  NativeCrypto.SSL_SESSION_up_ref(sslSessionNativePointer);
  this.sslSessionNativePointer = sslSessionNativePointer;
```

Though I guess that might be different from the rest of Conscrypt? Whatever works.